group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -544,6 +550,77 @@ public OffsetFetchResponseData.OffsetFetchResponseGroup fetchAllOffsets(
             .setTopics(topicResponses);
     }
 
+    /**
+     * Remove expired offsets for group.
+     *
+     * @param groupId The group id.
+     * @param records The list of records to populate with offset commit tombstone records.
+     *
+     * @return True if no offsets exist or if all offsets expired, false otherwise.
+     */
+    public boolean cleanupExpiredOffsets(String groupId, List<Record> records) {
+        TimelineHashMap<String, TimelineHashMap<Integer, OffsetAndMetadata>> offsetsByTopic = offsetsByGroup.get(groupId);
+        if (offsetsByTopic == null) {
+            return true;
+        }
+
+        // We expect the group to exist.
+        Group group = groupMetadataManager.group(groupId);
+        Set<String> expiredPartitions = new HashSet<>();
+        long currentTimestampMs = time.milliseconds();
+        Optional<OffsetExpirationCondition> offsetExpirationCondition = group.offsetExpirationCondition();
+
+        if (!offsetExpirationCondition.isPresent()) {
+            return false;
+        }
+
+        AtomicBoolean hasAllOffsetsExpired = new AtomicBoolean(true);
+        OffsetExpirationCondition condition = offsetExpirationCondition.get();
+
+        offsetsByTopic.forEach((topic, partitions) -> {
+            if (!group.isSubscribedToTopic(topic)) {
+                partitions.forEach((partition, offsetAndMetadata) -> {
+                    if (condition.isOffsetExpired(offsetAndMetadata, currentTimestampMs, config.offsetsRetentionMs)) {
+                        expiredPartitions.add(appendOffsetCommitTombstone(groupId, topic, partition, records).toString());
+                    } else {
+                        hasAllOffsetsExpired.set(false);
+                    }
+                });
+            } else {
+                hasAllOffsetsExpired.set(false);
+            }
+        });
+
+        if (!expiredPartitions.isEmpty()) {
+            log.info("[GroupId {}] Expiring offsets of partitions (hasAllOffsetsExpired={}): {}",
```

Comment:
Correct me if I'm wrong this is just for my understanding, so unless all the offsets are expired we don't delete the group right? In cases where hasAllOffsetsExpired is false and expiredPartitions is non-empty, the group won't be deleted but the tombstone records will be appended? And the next time we iterate through the partitions the ones with the tombstone record aren't included right?

Reply:
The placement of hasAllOffsetsExpired seems a bit off for the logging message `[GroupId 12345] Expiring offsets of partitions (hasAllOffsetsExpired=false): partition1, partition2, partition3` This is how it would look right? Can we move it to the end of the list or the beginning of this line if you think it makes sense as well